### PR TITLE
call _upcast to consider overflow for computing giou loss

### PR DESCRIPTION
**Pull request by developer0hye**
giou loss is weak at overflow problem because it computes area of box.

https://github.com/pytorch/vision/blob/main/torchvision/ops/boxes.py#L226-L242

### CODE REVIEW

File: torchvision/ops/giou_loss.py [hunk 1]
```diff
@@ -1,4 +1,13 @@
 import torch
+from torch import Tensor
+
+
+def _upcast(t: Tensor) -> Tensor:
+    # Protects from numerical overflows in multiplications by upcasting to the equivalent higher type
+    if t.is_floating_point():
+        return t if t.dtype in (torch.float32, torch.float64) else t.float()
+    else:
+        return t if t.dtype in (torch.int32, torch.int64) else t.int()
```
Comment (datumbox) [inline torchvision/ops/giou_loss.py:10]: This method already exists at [boxes.py](https://github.com/pytorch/vision/blob/289fce29b3e2392114aadbe7a419df0f2e3ac1be/torchvision/ops/boxes.py#L218-L223) but I'm not sure we should use it here. This is because it keeps boxes in integer float and that would break giou's `mean()` reduction. Instead we should cast to float if not already float32/float64.

**Review by datumbox — COMMENTED** [review]
@developer0hye Thanks for the PR. Looks like something we need to address. I've added a comment below for your attention, let me know your thoughts.

File: torchvision/ops/giou_loss.py (hunk 1 above)
Comment (developer0hye) [reply torchvision/ops/giou_loss.py:10]: @datumbox 

Thanks for your review!

Should we newly define upcast method for this function like the below code?

```python
if t.dtype not in (torch.float32, torch.float64):
 return t.float()
```

Comment (datumbox) [reply torchvision/ops/giou_loss.py:10]: A private method that does that would be fine. No strong opinions. :)

Comment (developer0hye) [reply torchvision/ops/giou_loss.py]: I modified it!

https://github.com/pytorch/vision/blob/dc2d754410e25557edc233c54b99b9f3247684fb/torchvision/ops/giou_loss.py#L5-L9

**Review by datumbox — APPROVED** [review]
LGTM, thanks!

**Comment by developer0hye** [comment]
@datumbox Thanks!
